### DISCUSSION TITLE
heif-convert calls error when importing(uploading) or converting heic/heif files with 'file not found'

### DISCUSSION THREAD

**Discussion by herroworrd** [photoprism/photoprism#2874] [Q&A] [answered]
#### 1. What is not working as documented?

When uploading (or converting) heif/heic files, heif-convert errors with file not found. Uploaded image will make it to originals folder but will be marked hidden in the index and unavailable in the UI.

#### 2. How can we reproduce it?

1. Upload an .heif image through Library > Import > Upload
2. Check resulting logs
3. Debug logs included below.

When reporting an import, indexing, or performance issue, please include the number and type of pictures in your library, as well as any configuration options you have changed, such as for thumbnail quality.

#### 3. What behavior do you expect?

Conversion of heic to jpeg to succeed 

#### 4. What could be the cause of your problem?
The files exist in /photoprism/originals folder and is in the index. I suppose a potential cause of this error is wrong inputs into the heif-convert command.

#### 5. Can you provide us with example files for testing, error logs, or screenshots?
```

time="2022-11-08T17:20:23-08:00" level=debug msg="upload: saving file 20210803_191856_DD4379D9.heif"
time="2022-11-08T17:20:23-08:00" level=info msg="1 files uploaded in 4 s"
time="2022-11-08T17:20:23-08:00" level=debug msg="server: POST /api/v1/users/uriaiuf1vmhvth4u/upload/k1k3x1 (200) [4.784969889s]"
time="2022-11-08T17:20:24-08:00" level=info msg="processing upload..."
time="2022-11-08T17:20:24-08:00" level=info msg="import: found no .ppignore file"
time="2022-11-08T17:20:24-08:00" level=info msg="import: added folder /assets/storage/users/uriaiuf1vmhvth4u/upload/sesssxfy5yczk1k3x1"
time="2022-11-08T17:20:24-08:00" level=info msg="media: 20210803_191856_DD4379D9.heif was taken at 2021-08-03 19:18:56 +0000 UTC (meta)"
time="2022-11-08T17:20:24-08:00" level=info msg="import: moving main heic file 20210803_191856_DD4379D9.heif to 2021/08/20210803_191856_DD4379D9.heic"
time="2022-11-08T17:20:24-08:00" level=debug msg="failed renaming file, fallback to copy and delete: rename /assets/storage/users/uriaiuf1vmhvth4u/uplo
ad/sesssxfy5yczk1k3x1/20210803_191856_DD4379D9.heif /photoprism/originals/2021/08/20210803_191856_DD4379D9.heic: invalid cross-device link"
time="2022-11-08T17:20:24-08:00" level=info msg="convert: converting 20210803_191856_DD4379D9.heic to 20210803_191856_DD4379D9.heic.jpg (heif-convert)"
time="2022-11-08T17:20:24-08:00" level=error msg="import: USAGE: /usr/bin/heif-convert [-q quality 0..100] <filename> <output>\nNo file specified\nWarn
ing: Error opening file - 92\nError: File not found - 92\n in 2021/08/20210803_191856_DD4379D9.heic (convert to jpeg)"
time="2022-11-08T17:20:24-08:00" level=debug msg="faces: found no orphan markers [16.466767ms]"
time="2022-11-08T17:20:24-08:00" level=debug msg="markers: found no invalid references [32.825363ms]"
time="2022-11-08T17:20:24-08:00" level=debug msg="markers: found no missing subjects [549.832µs]"
time="2022-11-08T17:21:44-08:00" level=debug msg="faces: found no ambiguous subjects [1m19.652619149s]"
time="2022-11-08T17:21:44-08:00" level=debug msg="faces: found no clusters to be merged [77.901802ms]"
time="2022-11-08T17:21:44-08:00" level=debug msg="faces: skipped clustering"
time="2022-11-08T17:21:44-08:00" level=debug msg="faces: found no new faces [6.553243ms]"
time="2022-11-08T17:21:44-08:00" level=debug msg="faces: found no unmatched markers"
time="2022-11-08T17:21:44-08:00" level=debug msg="faces: updated 0 markers, recognized 0 faces, 0 unknown [227.150013ms]"
time="2022-11-08T17:21:44-08:00" level=debug msg="index: updating counts"
time="2022-11-08T17:21:44-08:00" level=debug msg="counts: updated 0 places [54.283736ms]"
time="2022-11-08T17:21:44-08:00" level=debug msg="counts: updated 0 subjects [14.971966ms]"
time="2022-11-08T17:21:44-08:00" level=debug msg="counts: updated 0 labels [106.773351ms]"
time="2022-11-08T17:21:44-08:00" level=info msg="upload: deleted empty folder /assets/storage/users/uriaiuf1vmhvth4u/upload/sesssxfy5yczk1k3x1"
time="2022-11-08T17:21:44-08:00" level=info msg="upload: imported 2 files"
```

#### 6. Which software versions do you use?

(a) [PhotoPrism® CE Build [221105-7a295cab4]

(b) Database Type & Version: MariaDB

(c) Operating System Types & Versions: Linux

You can find the version/build number of the app in *Settings* by scrolling to the bottom. Note that MySQL 8 support has been discontinued, see system requirements at <https://docs.photoprism.app/getting-started/#system-requirements>.

*Always provide database and operating system details if it is a backend, import, or indexing issue. Should it be a frontend issue, at a minimum we require you to provide web browser and operating system details. When reporting a performance problem, we ask that you provide us with complete information about your environment, as there may be more than one cause.*

#### 7. On what kind of device is PhotoPrism installed?

(a) Device / Processor Type: TrueNas Scale (truecharts)

(b) Physical Memory & Swap Space in 16 GB

(c) Storage Type: SSD

(d) ~20,000 photos and only HEIC/HEIF are of issue

**Comment by lastzero** [comment, marked as answer]
Which of our Docker images do you use? Or could it be a custom version for TrueNAS? In that case please note that it is a third-party app and not maintained by us. You are welcome to test the HEIC support on our demo.

**Reply by herroworrd** [reply]
This is using 221105-bullseye. I'm assuming the demo site uses Ubuntu but it seems to work fine.

I wasn't aware you were maintaining multiple distros with each release and I'm not sure why truecharts isn't using Ubuntu now.

**Reply by lastzero** [reply]
Thanks for testing! The Debian images don't have full file format support, so we should document that somewhere / stop providing them. What is also strange is the assets path and that uploads go there as it may not be writable. Some of it could potentially be fixed by changing environment variables,  but not without any experience. It's easiest to use the Ubuntu image for now.

**Reply by lastzero** [reply]
Of course, fixing this would be a great task for contributors that want to use these images 👍

**Reply by herroworrd** [reply]
Maintaining multiple distros worth of docker images seems unnecessary and confusing. Especially if some of them aren't feature complete. 

I'm not associated with truecharts but I assume they're using bullseye only because they don't know any better. Truecharts doesn't target 'latest' so, the logical choice in their position, would be to continue the bullseye targets since this was the preferred(only?) base pre-1102.

**Reply by lastzero** [reply]
We don't recommended it and it's not part of our documentation. Why would you choose an undocumented, old Linux distribution if you don't know what it is good for? We used those for the previous release, but our release notes super clearly state Ubuntu is the default now and they even say why that is the case and what the advantages are.

**Reply by lastzero** [reply]
Maybe you could help communicating this while we delete those images knowing they are misused.

**Reply by lastzero** [reply]
Deleted. Note the TrueNAS version is completely broken now for new users and won't work anymore. Thanks for your help!

**Reply by lastzero** [reply]
Here's the announcement, spread the word:
> Note that we have deleted all Docker images for other distributions, as they are unnecessary and confusing. Especially because they are not feature complete.

**Reply by herroworrd** [reply]
Haha... great news! Having a hard time reading tone here but if you took offense to anything I've done or said, I'm sorry. I'd been excited for 1102 to release to get proraw support but was disappointed when more stuff broke.

Hopefully this change alleviates some workload on your end both in the form of dealing with stupid end-users, like me, questioning decisions but also now you only have to maintain one image?

I've submitted a PR to the truecharts team to update them of this situation. Cheers.

**Reply by lastzero** [reply]
Thanks for reporting this! Not sure for which reasons people use(d) our legacy builds. It can make sense, for example, when you don't need HEIC support and your host runs an old distribution so that hardware transcoding works. However, I'm glad we now have a great excuse not to provide extra images anymore! 🥳
